src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5570,36 +5511,40 @@ private static Task InternalWhenAll(Task[] tasks)
         // which involves several allocations, with this logic:
         //      return new WhenAllPromise(tasksCopy);
         // which saves a couple of allocations and enables debugger notification specialization.
-        //
-        // Used in InternalWhenAll(Task[])
         private sealed class WhenAllPromise : Task, ITaskCompletionAction
         {
-            /// <summary>
-            /// Stores all of the constituent tasks.  Tasks clear themselves out of this
-            /// array as they complete, but only if they don't have their wait notification bit set.
-            /// </summary>
-            private readonly Task?[] m_tasks;
             /// <summary>The number of tasks remaining to complete.</summary>
             private int m_count;
+            /// <summary>True if any of the supplied tasks require wait notification.</summary>
+            private bool m_anyTaskRequiresNotifyDebuggerOfWaitCompletion;
+            /// <summary>The tasks that faulted.</summary>
+            private ConcurrentQueue<Task>? m_faultedTasks;
+            /// <summary>The first task that got canceled.</summary>
+            private Task? m_canceledTask;
 
-            internal WhenAllPromise(Task[] tasks)
+            internal WhenAllPromise(IEnumerable<Task?>? tasks)
```

Comment:
Why `Task?` ?

Reply:
The null check for tasks is done in this constructor so the tasks passed could be null. But the tasks in the public API are not nullable. Not sure how it works.

Reply:
Ah, gotcha.  You can remove the inner `?`.  The signature is meant to suggest what the contract is, but the implementation can be more defensive in case the contract is violated.